File: src/services/authService.ts
// src/services/authService.ts

import api, { saveUser, clearUser, getCurrentUser } from './api';

// ============================================
// TYPES & INTERFACES
// ============================================

export interface User {
  id: string;
  name: string;
  fullName?: string;
  email: string;
  phone?: string;
  role?: string;
  roles?: string[];
  primaryRole?: string;
  upgrades?: string[];
  avatar?: string;
  profileImage?: string;
  bloodGroup?: string;
  gender?: string;
  token?: string;
  createdAt?: string;
  updatedAt?: string;
  isActive?: boolean;
  isVerified?: boolean;
  emailVerified?: boolean;
  phoneVerified?: boolean;
  isAdminApproved?: boolean;
  isOnline?: boolean;
  preferences?: UserPreferences;
}

export interface UserPreferences {
  notifications: {
    email: boolean;
    sms: boolean;
    push: boolean;
  };
  language: string;
  theme: string;
  timezone: string;
}

export interface LoginCredentials {
  email: string;
  password: string;
  role?: string;
  rememberMe?: boolean;
}

export interface RegisterData {
  name: string;
  full_name: string;
  email: string;
  phone: string;
  password: string;
  confirmPassword?: string;
  confirm_password?: string;
  role: string;
  gender?: string;
  acceptTerms?: boolean;
}

export interface AuthResponse {
  user: User;
  token: string;
  accessToken?: string;
  refreshToken?: string;
  expiresIn?: number;
}

export interface ChangePasswordData {
  currentPassword: string;
  current_password: string;
  newPassword: string;
  new_password: string;
  confirmNewPassword: string;
  confirm_new_password: string;
}

export interface ResetPasswordData {
  email?: string;
  token: string;
  newPassword: string;
  new_password: string;
  confirmNewPassword: string;
  confirm_password: string;
}

export interface ForgotPasswordResponse {
  message: string;
}

export interface VerifyEmailResponse {
  verified: boolean;
  message: string;
}

// ============================================
// AUTH SERVICE
// ============================================

export const authService = {
  /**
   * Login user with email and password
   */
  login: async (credentials: LoginCredentials): Promise<AuthResponse> => {
    const response = await api.post<any>('/auth/login', {
      email: credentials.email,
      password: credentials.password,
      role: credentials.role,
    });

    if (response.success && response.data) {
      const tokenData = response.data;
      const accessToken = tokenData.access_token;
      const refreshToken = tokenData.refresh_token;
      const userData = tokenData.user || {};

      // Normalize user object
      const user: User = {
        id: userData.id,
        name: userData.full_name || userData.name || credentials.email.split('@')[0],
        fullName: userData.full_name,
        email: userData.email || credentials.email,
        phone: userData.phone,
        role: userData.primary_role || userData.role || 'patient',
        roles: userData.roles || [userData.primary_role || 'patient'],
        primaryRole: userData.primary_role,
        upgrades: userData.upgrades || [],
        avatar: userData.profile_image,
        profileImage: userData.profile_image,
        bloodGroup: userData.blood_group,
        gender: userData.gender,
        token: accessToken,
        isVerified: userData.is_verified,
        emailVerified: userData.is_verified,
        isAdminApproved: userData.is_admin_approved,
        isActive: userData.is_active,
        isOnline: userData.is_online,
        createdAt: userData.created_at,
      };

      // Persist auth data
      saveUser({ ...user, token: accessToken });
      if (credentials.rememberMe && refreshToken) {
        localStorage.setItem('medicare_refresh_token', refreshToken);
      }

      return {
        user,
        token: accessToken,
        refreshToken,
        expiresIn: tokenData.expires_in,
      };
    }

    throw new Error(response.message || 'Login failed');
  },

  /**
   * Register new user
   */
  register: async (data: RegisterData): Promise<AuthResponse> => {
    const response = await api.post<any>('/auth/register', {
      full_name: data.full_name || data.name,
      email: data.email,
      phone: data.phone,
      password: data.password,
      confirm_password: data.confirm_password || data.confirmPassword,
      role: data.role,
      gender: data.gender,
    });

    if (response.success && response.data) {
      const tokenData = response.data;
      const accessToken = tokenData.access_token;
      const refreshToken = tokenData.refresh_token;
      const userData = tokenData.user || {};

      const user: User = {
        id: userData.id,
        name: userData.full_name || data.name,
        fullName: userData.full_name,
        email: userData.email || data.email,
        phone: userData.phone || data.phone,
        role: userData.primary_role || data.role,
        roles: userData.roles || [data.role],
        primaryRole: userData.primary_role,
        upgrades: userData.upgrades || [],
        avatar: userData.profile_image,
        profileImage: userData.profile_image,
        token: accessToken,
        isVerified: userData.is_verified || false,
        emailVerified: userData.is_verified || false,
        isActive: userData.is_active ?? true,
        isAdminApproved: userData.is_admin_approved || false,
        createdAt: userData.created_at,
      };

      saveUser({ ...user, token: accessToken });

      return {
        user,
        token: accessToken,
        refreshToken,
        expiresIn: tokenData.expires_in,
      };
    }

    throw new Error(response.message || 'Registration failed');
  },

  /**
   * Logout user
   */
  logout: async (): Promise<void> => {
    try {
      const refreshToken = localStorage.getItem('medicare_refresh_token');
      await api.post('/auth/logout', { refresh_token: refreshToken });
    } catch (error) {
      console.error('Logout error:', error);
    } finally {
      clearUser();
    }
  },

  /**
   * Get current authenticated user from localStorage
   */
  getCurrentUser: (): User | null => {
    return getCurrentUser() as User | null;
  },

  /**
   * Get user profile from server
   */
  getProfile: async (): Promise<User> => {
    const response = await api.get<any>('/auth/profile');

    if (response.success && response.data) {
      const userData = response.data;
      const currentUser = getCurrentUser() as User | null;
      const token = currentUser?.token;

      const user: User = {
        id: userData.id,
        name: userData.full_name || userData.name,
        fullName: userData.full_name,
        email: userData.email,
        phone: userData.phone,
        role: userData.primary_role || userData.role,
        roles: userData.roles,
        primaryRole: userData.primary_role,
        upgrades: userData.upgrades || [],
        avatar: userData.profile_image,
        profileImage: userData.profile_image,
        bloodGroup: userData.blood_group,
        gender: userData.gender,
        token,
        isVerified: userData.is_verified,
        emailVerified: userData.is_verified,
        isAdminApproved: userData.is_admin_approved,
        isActive: userData.is_active,
        isOnline: userData.is_online,
        createdAt: userData.created_at,
        updatedAt: userData.updated_at,
      };

      saveUser({ ...user, token });
      return user;
    }

    throw new Error(response.message || 'Failed to fetch profile');
  },

  /**
   * Update user profile
   */
  updateProfile: async (updates: Partial<User>): Promise<User> => {
    const response = await api.put<any>('/auth/profile', {
      full_name: updates.fullName || updates.name,
      phone: updates.phone,
      gender: updates.gender,
      profile_image: updates.avatar || updates.profileImage,
      blood_group: updates.bloodGroup,
    });

    if (response.success) {
      const currentUser = getCurrentUser() as User;
      const updatedUser = {
        ...currentUser,
        ...updates,
        updatedAt: new Date().toISOString(),
      };
      saveUser(updatedUser);
      return updatedUser;
    }

    throw new Error(response.message || 'Failed to update profile');
  },

  /**
   * Change user password
   */
  changePassword: async (data: ChangePasswordData): Promise<{ message: string }> => {
    const response = await api.post<any>('/auth/change-password', {
      current_password: data.currentPassword || data.current_password,
      new_password: data.newPassword || data.new_password,
      confirm_new_password: data.confirmNewPassword || data.confirm_new_password,
    });

    if (response.success) {
      return { message: response.message || 'Password changed successfully' };
    }

    throw new Error(response.message || 'Failed to change password');
  },

  /**
   * Request password reset
   */
  forgotPassword: async (email: string): Promise<ForgotPasswordResponse> => {
    const response = await api.post<any>('/auth/forgot-password', { email });

    if (response.success) {
      return { message: response.message || 'Password reset email sent.' };
    }

    throw new Error(response.message || 'Failed to request password reset');
  },

  /**
   * Reset password using token
   */
  resetPassword: async (data: ResetPasswordData): Promise<{ message: string }> => {
    const response = await api.post<any>('/auth/reset-password', {
      token: data.token,
      new_password: data.newPassword || data.new_password,
      confirm_password: data.confirmNewPassword || data.confirm_password,
    });

    if (response.success) {
      return { message: response.message || 'Password reset successfully' };
    }

    throw new Error(response.message || 'Failed to reset password');
  },

  /**
   * Verify email address
   */
  verifyEmail: async (token: string): Promise<VerifyEmailResponse> => {
    const response = await api.post<any>('/auth/verify-email', { token });

    if (response.success) {
      const currentUser = getCurrentUser() as User | null;
      if (currentUser) {
        const updatedUser = { ...currentUser, emailVerified: true, isVerified: true };
        saveUser(updatedUser);
      }
      return { verified: true, message: response.message || 'Email verified successfully' };
    }

    throw new Error(response.message || 'Email verification failed');
  },

  /**
   * Resend verification email
   */
  resendVerificationEmail: async (email?: string): Promise<{ message: string }> => {
    const response = await api.post<any>('/auth/verify-email', { email });

    return { message: response.message || 'Verification email sent' };
  },

  /**
   * Refresh authentication token
   */
  refreshToken: async (): Promise<{ token: string; refreshToken?: string }> => {
    const refreshToken = localStorage.getItem('medicare_refresh_token');
    if (!refreshToken) {
      throw new Error('No refresh token available');
    }

    const response = await api.post<any>('/auth/refresh', {
      refresh_token: refreshToken,
    });

    if (response.success && response.data) {
      const newToken = response.data.access_token;
      const newRefreshToken = response.data.refresh_token;
      const currentUser = getCurrentUser() as User | null;

      if (currentUser) {
        const userData = { ...currentUser, token: newToken };
        saveUser(userData);
        if (newRefreshToken) {
          localStorage.setItem('medicare_refresh_token', newRefreshToken);
        }
      }

      return { token: newToken, refreshToken: newRefreshToken };
    }

    throw new Error(response.message || 'Token refresh failed');
  },

  /**
   * Switch active role
   */
  switchRole: async (role: string): Promise<{ token: string; role: string }> => {
    const response = await api.post<any>('/auth/switch-role', { role });

    if (response.success && response.data) {
      const newToken = response.data.access_token;
      const currentUser = getCurrentUser() as User | null;

      if (currentUser) {
        const userData = {
          ...currentUser,
          token: newToken,
          primaryRole: role,
          role,
        };
        saveUser(userData);
      }

      return { token: newToken, role };
    }

    throw new Error(response.message || 'Role switch failed');
  },

  /**
   * Check if user is authenticated
   */
  isAuthenticated: (): boolean => {
    const user = authService.getCurrentUser();
    return !!user && !!user.token;
  },

  /**
   * Check if user has specific role
   */
  hasRole: (roles: string | string[]): boolean => {
    const user = authService.getCurrentUser();
    if (!user) return false;

    const roleList = Array.isArray(roles) ? roles : [roles];
    const userRoles = user.roles || [user.role] || [];
    return roleList.some(r => userRoles.includes(r) || user.primaryRole === r || user.role === r);
  },

  /**
   * Check if email is verified
   */
  isEmailVerified: (): boolean => {
    const user = authService.getCurrentUser();
    return user?.emailVerified || user?.isVerified || false;
  },

  /**
   * Get user preferences
   */
  getUserPreferences: (): UserPreferences | null => {
    const user = authService.getCurrentUser();
    return user?.preferences || null;
  },

  /**
   * Update user preferences
   */
  updatePreferences: async (preferences: Partial<UserPreferences>): Promise<UserPreferences> => {
    const currentUser = authService.getCurrentUser();
    if (!currentUser) {
      throw new Error('Not authenticated');
    }

    const updatedPreferences: UserPreferences = {
      ...(currentUser.preferences ?? {
        notifications: { email: true, sms: true, push: true },
        language: 'en',
        theme: 'system',
        timezone: 'UTC',
      }),
      ...preferences,
    };

    await authService.updateProfile({ preferences: updatedPreferences });
    return updatedPreferences;
  },
};

// ============================================
// AUTH HOOK
// ============================================

export const useAuth = () => {
  return {
    user: authService.getCurrentUser(),
    isAuthenticated: authService.isAuthenticated(),
    isAdmin: authService.hasRole(['admin', 'super_admin']),
    isDoctor: authService.hasRole('doctor'),
    isPatient: authService.hasRole(['patient', 'client']),
    isEmailVerified: authService.isEmailVerified(),
    login: authService.login,
    register: authService.register,
    logout: authService.logout,
    updateProfile: authService.updateProfile,
    changePassword: authService.changePassword,
    refreshToken: authService.refreshToken,
    switchRole: authService.switchRole,
  };
};

export default authService;

File: backend/app/services/auth_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException
from datetime import datetime, timedelta
from typing import Optional

from ..models.user import (
    User, UserProfile, UserSession, UserRole, Role, PasswordReset,
)
from ..schemas.auth import (
    LoginRequest, RegisterRequest, TokenResponse, UserBasicResponse,
    RefreshTokenResponse,
)
from ..core.security import (
    hash_password, verify_password, create_access_token, create_refresh_token, decode_token,
)
from ..core.exceptions import (
    UnauthorizedException, ConflictException, BadRequestException, NotFoundException,
)
from ..core.config import get_settings

settings = get_settings()


class AuthService:
    """Handles all authentication operations with real database queries."""

    def __init__(self, db: Session):
        self.db = db

    # ============================================
    # LOGIN
    # ============================================
    async def login(self, request: LoginRequest) -> TokenResponse:
        user = self.db.query(User).filter(User.email == request.email).first()
        if not user:
            raise UnauthorizedException("Invalid email or password")

        if not user.is_active:
            raise UnauthorizedException("Account is deactivated")

        if not verify_password(request.password, user.password_hash):
            # Track failed attempt via security log
            raise UnauthorizedException("Invalid email or password")

        # Load user roles
        roles = self._get_user_roles(user.id)
        primary_role = self._get_primary_role_name(user.primary_role_id)

        # Create tokens
        token_data = {"sub": user.id, "role": primary_role, "roles": roles}
        access_token = create_access_token(token_data)
        refresh_token = create_refresh_token({"sub": user.id})

        # Store refresh token in user_sessions
        self._create_session(user.id, refresh_token)

        # Update online status
        user.is_online = True
        self.db.commit()

        user_response = self._build_user_response(user, roles, primary_role)
        return TokenResponse(
            access_token=access_token,
            refresh_token=refresh_token,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
            user=user_response,
        )

    # ============================================
    # REGISTER
    # ============================================
    async def register(self, request: RegisterRequest) -> TokenResponse:
        if request.password != request.confirm_password:
            raise BadRequestException("Passwords do not match")

        # Check if email already exists
        existing = self.db.query(User).filter(User.email == request.email).first()
        if existing:
            raise ConflictException("An account with this email already exists")

        # Resolve role
        role_name = request.role or "patient"
        role = self.db.query(Role).filter(Role.name == role_name).first()
        if not role:
            # Fallback to client role if specific role not found
            role = self.db.query(Role).filter(Role.name == "client").first()
        if not role:
            role = self.db.query(Role).first()  # Absolute fallback

        # Create user
        user = User(
            email=request.email,
            full_name=request.full_name,
            phone=request.phone,
            password_hash=hash_password(request.password),
            gender=request.gender,
            primary_role_id=role.id if role else "",
            is_verified=False,
            is_active=True,
            is_admin_approved=False,
        )
        self.db.add(user)
        self.db.flush()  # Get user.id

        # Create user profile
        profile = UserProfile(
            user_id=user.id,
            first_name=request.full_name.split(" ")[0] if request.full_name else "",
            last_name=" ".join(request.full_name.split(" ")[1:]) if request.full_name and " " in request.full_name else "",
        )
        self.db.add(profile)

        # Assign role
        if role:
            user_role = UserRole(user_id=user.id, role_id=role.id)
            self.db.add(user_role)

        self.db.commit()
        self.db.refresh(user)

        # Auto-login after registration
        roles = self._get_user_roles(user.id)
        primary_role = self._get_primary_role_name(user.primary_role_id)

        token_data = {"sub": user.id, "role": primary_role, "roles": roles}
        access_token = create_access_token(token_data)
        refresh_token = create_refresh_token({"sub": user.id})

        self._create_session(user.id, refresh_token)
        self.db.commit()

        user_response = self._build_user_response(user, roles, primary_role)
        return TokenResponse(
            access_token=access_token,
            refresh_token=refresh_token,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
            user=user_response,
        )

    # ============================================
    # REFRESH TOKEN (with rotation)
    # ============================================
    async def refresh_token(self, refresh_token_str: str) -> RefreshTokenResponse:
        payload = decode_token(refresh_token_str)
        if not payload or payload.get("type") != "refresh":
            raise UnauthorizedException("Invalid refresh token")

        user_id = payload.get("sub")
        if not user_id:
            raise UnauthorizedException("Invalid token payload")

        # Verify session exists and not revoked
        session = (
            self.db.query(UserSession)
            .filter(
                UserSession.refresh_token == refresh_token_str,
                UserSession.user_id == user_id,
                UserSession.is_revoked == False,
                UserSession.expires_at > datetime.utcnow(),
            )
            .first()
        )
        if not session:
            raise UnauthorizedException("Refresh token expired or revoked")

        # Verify user still active
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user or not user.is_active:
            raise UnauthorizedException("User account is deactivated")

        # ── Token rotation: revoke old refresh token ──
        session.is_revoked = True

        # Create new token pair
        roles = self._get_user_roles(user.id)
        primary_role = self._get_primary_role_name(user.primary_role_id)

        token_data = {"sub": user.id, "role": primary_role, "roles": roles}
        new_access_token = create_access_token(token_data)
        new_refresh_token = create_refresh_token({"sub": user.id})

        # Store new refresh token session
        self._create_session(user.id, new_refresh_token)
        self.db.commit()

        return RefreshTokenResponse(
            access_token=new_access_token,
            refresh_token=new_refresh_token,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
        )

    # ============================================
    # LOGOUT
    # ============================================
    async def logout(self, user_id: str, refresh_token: Optional[str] = None) -> None:
        user = self.db.query(User).filter(User.id == user_id).first()
        if user:
            user.is_online = False

        if refresh_token:
            self.db.query(UserSession).filter(
                UserSession.refresh_token == refresh_token,
                UserSession.user_id == user_id,
            ).update({"is_revoked": True})
        else:
            # Revoke all sessions for this user
            self.db.query(UserSession).filter(
                UserSession.user_id == user_id,
            ).update({"is_revoked": True})

        self.db.commit()

    # ============================================
    # FORGOT PASSWORD
    # ============================================
    async def forgot_password(self, email: str) -> str:
        user = self.db.query(User).filter(User.email == email).first()
        if not user:
            # Don't reveal whether email exists — return success anyway
            return "If an account with this email exists, a reset link has been sent."

        # Generate reset token
        import uuid
        token = str(uuid.uuid4())
        expires_at = datetime.utcnow() + timedelta(hours=1)

        password_reset = PasswordReset(
            user_id=user.id,
            token=token,
            expires_at=expires_at,
        )
        self.db.add(password_reset)
        self.db.commit()

        # TODO: Send actual email with reset link
        # reset_url = settings.PASSWORD_RESET_URL.format(
        #     frontend_url=settings.FRONTEND_URL, token=token
        # )
        # await send_email(user.email, "Password Reset", f"Click: {reset_url}")

        return "If an account with this email exists, a reset link has been sent."

    # ============================================
    # RESET PASSWORD
    # ============================================
    async def reset_password(self, token: str, new_password: str) -> None:
        reset_record = (
            self.db.query(PasswordReset)
            .filter(
                PasswordReset.token == token,
                PasswordReset.used_at.is_(None),
                PasswordReset.expires_at > datetime.utcnow(),
            )
            .first()
        )
        if not reset_record:
            raise BadRequestException("Invalid or expired reset token")

        user = self.db.query(User).filter(User.id == reset_record.user_id).first()
        if not user:
            raise NotFoundException("User not found")

        user.password_hash = hash_password(new_password)
        reset_record.used_at = datetime.utcnow()
        self.db.commit()

    # ============================================
    # CHANGE PASSWORD
    # ============================================
    async def change_password(self, user_id: str, current_password: str, new_password: str) -> None:
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            raise NotFoundException("User not found")

        if not verify_password(current_password, user.password_hash):
            raise BadRequestException("Current password is incorrect")

        user.password_hash = hash_password(new_password)
        self.db.commit()

    # ============================================
    # VERIFY EMAIL
    # ============================================
    async def verify_email(self, token: str) -> None:
        # Token format: "verify-{user_id}"
        # In production, use a proper email verification token table
        if token.startswith("verify-"):
            user_id = token.replace("verify-", "")
            user = self.db.query(User).filter(User.id == user_id).first()
            if user:
                user.is_verified = True
                self.db.commit()
        else:
            raise BadRequestException("Invalid verification token")

    # ============================================
    # HELPER METHODS
    # ============================================
    def _get_user_roles(self, user_id: str) -> list[str]:
        role_rows = (
            self.db.query(Role.name)
            .join(UserRole, UserRole.role_id == Role.id)
            .filter(UserRole.user_id == user_id, UserRole.is_active == True, Role.is_active == True)
            .all()
        )
        return [r[0] for r in role_rows]

    def _get_primary_role_name(self, primary_role_id: str) -> str:
        if not primary_role_id:
            return "client"
        role = self.db.query(Role).filter(Role.id == primary_role_id).first()
        return role.name if role else "client"

    def _create_session(self, user_id: str, refresh_token: str) -> None:
        session = UserSession(
            user_id=user_id,
            refresh_token=refresh_token,
            expires_at=datetime.utcnow() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS),
        )
        self.db.add(session)

    def _build_user_response(self, user: User, roles: list[str], primary_role: str) -> UserBasicResponse:
        return UserBasicResponse(
            id=user.id,
            email=user.email,
            full_name=user.full_name,
            phone=user.phone,
            gender=user.gender,
            profile_image=user.profile_image,
            blood_group=user.blood_group,
            primary_role=primary_role,
            roles=roles,
            upgrades=[],  # TODO: load from user_role_upgrades
            is_verified=user.is_verified,
            is_admin_approved=user.is_admin_approved,
            is_active=user.is_active,
            is_online=user.is_online,
            created_at=user.created_at,
        )
